connector/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java

```diff
@@ -0,0 +1,121 @@
+package com.google.cloud.spark.bigquery;
+
+import com.google.protobuf.ByteString;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.ipc.ArrowReader;
+import org.apache.arrow.vector.ipc.ArrowStreamReader;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.vectorized.ColumnVector;
+import org.apache.spark.sql.vectorized.ColumnarBatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.SequenceInputStream;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class ArrowBinaryIterator implements Iterator<InternalRow> {
+
+    private static long maxAllocation = Long.MAX_VALUE;
+    ArrowReaderIterator arrowReaderIterator;
+    Iterator<InternalRow> currentIterator;
+    List<String> columnsInOrder;
+
+    public ArrowBinaryIterator(List<String> columnsInOrder, ByteString schema, ByteString rowsInBytes) {
+        BufferAllocator allocator = (new RootAllocator(maxAllocation)).newChildAllocator("ArrowBinaryIterator",
+                0, maxAllocation);
+
+        SequenceInputStream bytesWithSchemaStream = new SequenceInputStream(
+                new ByteArrayInputStream(schema.toByteArray()),
+                new ByteArrayInputStream(rowsInBytes.toByteArray()));
+
+        ArrowStreamReader arrowStreamReader = new ArrowStreamReader(bytesWithSchemaStream, allocator);
+        arrowReaderIterator = new ArrowReaderIterator(arrowStreamReader);
+        this.columnsInOrder = columnsInOrder;
+        initializeCurrentIterator();
+    }
+
+    @Override
+    public boolean hasNext() {
+        initializeCurrentIterator();
+        return currentIterator.hasNext();
```

Comment:
You can simplify the implementation:
```
public boolean hasNext() {
        while(!currentIterator.hasNext()) {
            if(! arrowReaderIterator.hasNext()) {
                return false;
            }
            currentIterator = toArrowRows(arrowReaderIterator.next(), columnsInOrder);
        }
        return currentIterator.hasNext();
}

Reply:
I am thinking of keeping a separate initialize function for refilling current iterator. Is that okay?